src/ECER.Resources.Documents/Certifications/CertificationRepository.cs

```diff
@@ -19,21 +19,54 @@ public async Task<IEnumerable<Certification>> Query(UserCertificationQuery query
   {
     await Task.CompletedTask;
     var Certifications = context.ecer_CertificateSet.AsQueryable();
+    var Registrants = context.ContactSet.AsQueryable();
 
     // Filtering by ID
     if (query.ById != null) Certifications = Certifications.Where(r => r.ecer_CertificateId == Guid.Parse(query.ById));
 
     // Filtering by applicant ID
     if (query.ByApplicantId != null) Certifications = Certifications.Where(r => r.ecer_Registrantid.Id == Guid.Parse(query.ByApplicantId));
 
-    Certifications = Certifications.OrderByDescending(r => r.ecer_ExpiryDate);
+    // Filtering by First Name
+    if (!string.IsNullOrEmpty(query.ByFirstName))
+    {
+      Certifications = from cert in Certifications
+                       join reg in Registrants on cert.ecer_Registrantid.Id equals reg.Id
+                       where reg.FirstName == (query.ByFirstName)
+                       select cert;
+    }
+
+    // Filtering by Last Name
+    if (!string.IsNullOrEmpty(query.ByLastName))
+    {
+      Certifications = from cert in Certifications
+                       join reg in Registrants on cert.ecer_Registrantid.Id equals reg.Id
+                       where reg.FirstName == (query.ByLastName)
+                       select cert;
+    }
+
+    // Filtering by certificate number
+    if (!string.IsNullOrEmpty(query.ByCertificateNumber))
+      Certifications = Certifications.Where(r => r.ecer_CertificateNumber.Contains(query.ByCertificateNumber));
 
     var results = context.From(Certifications)
       .Join()
       .Include(a => a.ecer_certifiedlevel_CertificateId)
       .Include(a => a.ecer_documenturl_CertificateId)
-      .Execute();
+      .Include(a => a.ecer_certificate_Registrantid)
+      .IncludeNested(a => a.ecer_certificateconditions_Registrantid)
+      .Execute().GroupBy(r => r.ecer_Registrantid.Id) // Group by unique identifier (assuming RegistrantId)
+             .Select(g => g.OrderByDescending(r => r.ecer_ExpiryDate).FirstOrDefault()); // Select latest by expiry date
+
+    // Calculate the correct number of items to skip
+    int skipAmount = 0;
+    if (query.PageNumber > 1)
+    {
+      skipAmount = (query.PageNumber - 1) * query.PageSize;
+    }
+
+    var paginatedResults = results.Skip(skipAmount).Take(query.PageSize); // Apply Pagination
```

Comment:
Pagination with the newly implemented method is not working like this, I believe you should put the page number directly in the Skip function. see communication repository
`      communications = communications.OrderByDescending(item => item.ecer_LatestMessageNotifiedDate).Skip(query.PageNumber).Take(query.PageSize);
`

Reply:
Here, You first get all certificates from Dataverse. when you call Execute() then you are working on results you got. this might cause performance issues as certificates grows.

Reply:
We have decided to revisit this issue in the future, as the Dataverse SDK currently lacks GroupBy support, and implementing it now would be too time-consuming and is not our priority. In the meantime, we’ve opted to handle it in memory, as the expected result size is manageable given the filtering by first name, last name, and other fields.